engine/src/core/graphics/rendering/rendering_device/batch_2d.cpp

```diff
@@ -0,0 +1,158 @@
+        if(m_UsedVertices + SPRITE_SIZE > MAX_VERTICES) return false; // TODO: Error handling using expected!
+
+        const auto textureID = sprite.GetTexture()->GetID();
+
+        float textureSlot = 0.0f;
+        if(textureID > 0) {
+            bool found = false;
+            for(std::size_t i = 0; i < m_Textures.size(); i++) {
+                if(m_Textures[i] == textureID) {
+                    found = true;
+                    textureSlot = static_cast<float>(i);
+                    break;
+                }
+            }
+
+            if(!found) {
+                if(m_Textures.size() >= MAX_TEXTURES_PER_SHADER) {
+                    return false;
+                }
+                m_Textures.push_back(textureID);
+                textureSlot = static_cast<float>(m_Textures.size() - 1);
+            }
+        }
+
+        const auto& position = sprite.GetPosition();
+        const auto& size = sprite.GetSize();
+        const auto& color = sprite.GetColor();
+
+        m_MappedBuffer->m_Position = {position.x, position.y, position.z};
+        m_MappedBuffer->m_Color = color;
+        m_MappedBuffer->m_TextureCoord = {0.0f, 0.0f};
+        m_MappedBuffer->m_TextureId = textureSlot;
+        ++m_MappedBuffer;
+
+        m_MappedBuffer->m_Position = {position.x, position.y + size.y, position.z};
+        m_MappedBuffer->m_Color = color;
+        m_MappedBuffer->m_TextureCoord = {0.0f, 1.0f};
+        m_MappedBuffer->m_TextureId = textureSlot;
+        ++m_MappedBuffer;
+
+        m_MappedBuffer->m_Position = {position.x + size.x, position.y + size.y, position.z};
+        m_MappedBuffer->m_Color = color;
+        m_MappedBuffer->m_TextureCoord = {1.0f, 1.0f};
+        m_MappedBuffer->m_TextureId = textureSlot;
+        ++m_MappedBuffer;
+
+        m_MappedBuffer->m_Position = {position.x + size.x, position.y, position.z};
+        m_MappedBuffer->m_Color = color;
+        m_MappedBuffer->m_TextureCoord = {1.0f, 0.0f};
+        m_MappedBuffer->m_TextureId = textureSlot;
+        ++m_MappedBuffer;
+
+        m_UsedVertices += SPRITE_SIZE;
+        m_UsedIndices += INDICES_SIZE;
+
+        return true;
+    }
+
+    void Batch2D::Begin(const Camera2D& camera) noexcept {
+        m_QuadShader->Bind();
+        m_QuadShader->SetUniformMatrix4fv("u_ProjectionView", camera.GetProjectionViewMatrix());
+        m_QuadShader->Unbind();
+
+        m_QuadVBO->Bind();
+        m_MappedBuffer = reinterpret_cast<SpriteVertexData*>(glMapBuffer(GL_ARRAY_BUFFER, GL_WRITE_ONLY));
+    }
+
+    void Batch2D::End() const noexcept {
+        glUnmapBuffer(GL_ARRAY_BUFFER);
+        m_QuadVBO->Unbind();
+    }
+
+    void Batch2D::Draw() noexcept {
+        m_QuadVAO->Bind();
+        m_QuadShader->Bind();
+        m_QuadShader->SetUniformMatrix4fv("u_Model", glm::mat4(1.0f));
+        for(std::size_t i = 0; i < m_Textures.size(); i++) {
+            glActiveTexture(GL_TEXTURE0 + static_cast<uint32_t>(i));
```

Comment:
why not just declare "i" as uint32_t?

Reply:
There is much confusion with types from the GLAD side. This is generally a messy API. We won't mitigate these stupid casts unless we want warnings.

Reply:
Also, this static_cast is the leftover from my previous testings (it's useless here, cause std::size_t is indeed uint32_t). I will change it.